### DISCUSSION TITLE
API to perform an event upon WS disconnect

### DISCUSSION THREAD

**Discussion by Archmonger** [idom-team/idom#528] [Ideas]
There needs to be some API to perform actions on disconnect.

Although this can currently be done by writing custom client websockets with listeners `onclose`, it isn't very convenient.

Events may be things such as:
- Replacing the component's content with static pre-rendered HTML (ex. loading animation)
- Executing pre-written JavaScript (this might be difficult to implement without expanding our the IDOM JavaScript API)

**Comment by rmorshea** [comment]
So I think the cleanest way to do this would probably be to allow all elements to have a `fallback`. Then, in the event of a websocket disconnection the root most elements of the layout that currently have fallbacks would display them. Not surprisingly, the fallback would not be allowed to have state or respond to events. This would unify things a bit actually since all elements would have a concept of a fallback rather than just imported web modules.

Implementing this in the JS client will be a bit messy but it should be doable.

**Reply by Archmonger** [reply]
Fallback alone works for my purposes.

Advanced users might want a JS API to more easily latch onto a view ID's WS events. But personally I don't think I'll need that, and I can write a custom WS if comes down to that.
